src/twisted/test/test_iosim.py

```diff
@@ -39,3 +44,85 @@ def test_writeSequence(self):
         a.writeSequence([b"b", b"c", b"d"])
 
         self.assertEqual(b"".join(a.stream), b"abcd")
+
+
+
+@implementer(IPushProducer)
+class StrictPushProducer(object):
```

Comment:
If you'd like to know the state just to make errors more informative, Automat now does that:

```python
from automat import MethodicalMachine
from twisted.internet.interfaces import IPushProducer
from zope.interface import implementer


def _returnNone(outputs):
    list(outputs)


@implementer(IPushProducer)
class StrictPushProducer(object):
    _machine = MethodicalMachine()

    @_machine.state(initial=True)
    def running(self):
        pass

    @_machine.state()
    def stopped(self):
        pass

    @_machine.state()
    def paused(self):
        pass

    @_machine.input()
    def stopProducing(self):
        pass

    @_machine.input()
    def pauseProducing(self):
        pass

    @_machine.input()
    def resumeProducing(self):
        pass

    running.upon(pauseProducing, enter=paused, outputs=[],
                 collector=_returnNone)
    running.upon(stopProducing, enter=stopped, outputs=[],
                 collector=_returnNone)
    paused.upon(resumeProducing, enter=running, outputs=[],
                collector=_returnNone)
    paused.upon(stopProducing, enter=stopped, outputs=[],
                collector=_returnNone)


validTransitions = StrictPushProducer()
validTransitions.pauseProducing()
validTransitions.resumeProducing()
validTransitions.stopProducing()

invalidTransition = StrictPushProducer()
invalidTransition.resumeProducing()
```

`invalidTransition` will fail like so:
```
Traceback (most recent call last):
  File "/tmp/machine.py", line 54, in <module>
    invalidTransition.resumeProducing()
  File "/home/mrw/src/python/automat/automat/_methodical.py", line 97, in resumeProducing
    outputs = transitioner.transition(self)
  File "/home/mrw/src/python/automat/automat/_core.py", line 147, in transition
    inputSymbol)
  File "/home/mrw/src/python/automat/automat/_core.py", line 129, in outputForInput
    raise NoTransition(state=inState, symbol=inputSymbol)
automat._core.NoTransition: no transition for MethodicalInput(method=<function resumeProducing at 0x7fcb3ef65938>) in MethodicalState(method=<function running at 0x7fcb3ef65500>)
```

The reporting could be improved but I believe it provides the same information the current implementation of `StrictPushProducer` does.